# Investigating, cleaning, and wrangling our dataset
1. Importing libraries and datasets
2. Exploratory data analysis 
3. Addressing our new data set
    1. Wrangling
    2. Consistency checks 
4. Merging population data
    1. Add in population for each location by merging new dataset 
    2. Creating population flag
    3. Creating job grouping 
    4. Creating age grouping
5. Converting categorical variables to numeric 
    1. Variables with two distinct values
          1. Gender
          2. Employment status
          3. Business unit
    2. Variables with multiple distinct values
        1. City name
        2. Department
        3. Term reason
        4. Term type
        5. Population flag
        6. Seniority
        7. Age flag
    3. Creating the encoded dataframe
6. Exporting dataset

# 1. Importing libraries and datasets 

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder 

In [2]:
#create path
path = r'C:\Users\rutha\CareerFoundry\04-23_EmployeeAttrition'

In [3]:
#import data set
df = pd.read_csv(os.path.join(path, '02_data', 'Original_data', 'EU10YearTermination.csv'), index_col=False)

In [4]:
#checking output
print('Sample of attrition data set')
df.sample(10)

Sample of attrition data set


EmployeeID   recorddate_key birthdate_key orighiredate_key  \
16116        3972  12/31/2009 0:00     3/15/1963        2/25/1997   
1904         1767  12/31/2012 0:00    06/02/1951       12/06/1990   
39844        6749  12/31/2012 0:00    10/09/1982        6/28/2007   
29274        5621  12/31/2011 0:00    08/07/1973        8/24/2002   
26670        5342  12/31/2011 0:00    06/11/1971       07/04/2001   
46912        7925  12/31/2012 0:00     9/27/1991        3/22/2012   
30591        5760  12/31/2009 0:00     9/30/1974       04/02/2003   
21728        4754  12/31/2008 0:00    08/08/1967        6/24/1999   
17532        4180  12/31/2009 0:00    05/09/1964       10/05/1997   
30113        5710  12/31/2013 0:00    05/06/1974        1/14/2003   

      terminationdate_key  age  length_of_service         country  \
16116          01/01/1900   46                 12         Denmark   
1904           01/01/1900   61                 22         Germany   
39844          01/01/1900   30                  5  United Kingdom   
29274          01/01/1900   38                  9         Hungary   
26670          01/01/1900   40                 10     Netherlands   
46912          01/01/1900   21                  0           Italy   
30591          01/01/1900   35                  6           Italy   
21728          01/01/1900   41                  9         Germany   
17532          01/01/1900   45                 12        Slovenia   
30113          01/01/1900   39                 10          Norway   

        department_name       job_title  store_name gender_short gender_full  \
16116           Produce   Produce Clerk          17            F      Female   
1904             Bakery  Bakery Manager          44            M        Male   
39844   Processed Foods   Shelf Stocker          31            F      Female   
29274             Dairy    Dairy Person          26            F      Female   
26670             Dairy    Dairy Person          18            F      Female   
46912  Customer Service         Cashier          21            M        Male   
30591             Dairy    Dairy Person          21            M        Male   
21728   Processed Foods   Shelf Stocker          42            M        Male   
17532           Produce   Produce Clerk           2            F      Female   
30113            Bakery           Baker           8            M        Male   

      termreason_desc   termtype_desc  STATUS_YEAR  STATUS BUSINESS_UNIT  
16116  Not Applicable  Not Applicable         2009  ACTIVE        STORES  
1904   Not Applicable  Not Applicable         2012  ACTIVE        STORES  
39844  Not Applicable  Not Applicable         2012  ACTIVE        STORES  
29274  Not Applicable  Not Applicable         2011  ACTIVE        STORES  
26670  Not Applicable  Not Applicable         2011  ACTIVE        STORES  
46912  Not Applicable  Not Applicable         2012  ACTIVE        STORES  
30591  Not Applicable  Not Applicable         2009  ACTIVE        STORES  
21728  Not Applicable  Not Applicable         2008  ACTIVE        STORES  
17532  Not Applicable  Not Applicable         2009  ACTIVE        STORES  
30113  Not Applicable  Not Applicable         2013  ACTIVE        STORES

In [5]:
print('Number of rows and columns in data set')
df.shape

Number of rows and columns in data set


(49653, 18)

**Initial Observations**

The dataset is 49,653 rows and has 18 columns. The variables cover things such as country location, department, gender, and whether or not they are still employed at our company.

# 2. Exploratory Data Analysis

Conducting some initial exploratory analysis to understand the data frame. 

In [6]:
df.head(5)

EmployeeID   recorddate_key birthdate_key orighiredate_key  \
0        1318  12/31/2006 0:00    01/03/1954        8/28/1989   
1        1318  12/31/2007 0:00    01/03/1954        8/28/1989   
2        1318  12/31/2008 0:00    01/03/1954        8/28/1989   
3        1318  12/31/2009 0:00    01/03/1954        8/28/1989   
4        1318  12/31/2010 0:00    01/03/1954        8/28/1989   

  terminationdate_key  age  length_of_service  country department_name  \
0          01/01/1900   52                 17  Germany       Executive   
1          01/01/1900   53                 18  Germany       Executive   
2          01/01/1900   54                 19  Germany       Executive   
3          01/01/1900   55                 20  Germany       Executive   
4          01/01/1900   56                 21  Germany       Executive   

  job_title  store_name gender_short gender_full termreason_desc  \
0       CEO          35            M        Male  Not Applicable   
1       CEO          35            M        Male  Not Applicable   
2       CEO          35            M        Male  Not Applicable   
3       CEO          35            M        Male  Not Applicable   
4       CEO          35            M        Male  Not Applicable   

    termtype_desc  STATUS_YEAR  STATUS BUSINESS_UNIT  
0  Not Applicable         2006  ACTIVE    HEADOFFICE  
1  Not Applicable         2007  ACTIVE    HEADOFFICE  
2  Not Applicable         2008  ACTIVE    HEADOFFICE  
3  Not Applicable         2009  ACTIVE    HEADOFFICE  
4  Not Applicable         2010  ACTIVE    HEADOFFICE

In [7]:
df.tail(5)

EmployeeID    recorddate_key birthdate_key orighiredate_key  \
49648        8258  12/01/2015 00:00     5/28/1994        8/19/2013   
49649        8264  08/01/2013 00:00     6/13/1994        8/27/2013   
49650        8279  12/01/2015 00:00     7/18/1994        9/15/2013   
49651        8296  12/01/2013 00:00    09/02/1994       10/09/2013   
49652        8321  12/01/2014 00:00    11/28/1994       11/24/2013   

      terminationdate_key  age  length_of_service   country   department_name  \
49648          12/30/2015   21                  2   Andorra             Dairy   
49649           8/30/2013   19                  0   Germany  Customer Service   
49650          12/30/2015   21                  2  Slovakia  Customer Service   
49651          12/31/2013   19                  0   Romania  Customer Service   
49652          12/30/2014   20                  1    Greece  Customer Service   

          job_title  store_name gender_short gender_full termreason_desc  \
49648  Dairy Person          34            M        Male          Layoff   
49649       Cashier          44            F      Female      Resignaton   
49650       Cashier          39            F      Female          Layoff   
49651       Cashier          16            F      Female      Resignaton   
49652       Cashier          13            F      Female          Layoff   

      termtype_desc  STATUS_YEAR      STATUS BUSINESS_UNIT  
49648   Involuntary         2015  TERMINATED        STORES  
49649     Voluntary         2013  TERMINATED        STORES  
49650   Involuntary         2015  TERMINATED        STORES  
49651     Voluntary         2013  TERMINATED        STORES  
49652   Involuntary         2014  TERMINATED        STORES

In [8]:
df.describe()

EmployeeID           age  length_of_service    store_name  \
count  49653.000000  49653.000000       49653.000000  49653.000000   
mean    4859.495740     42.077035          10.434596     27.297605   
std     1826.571142     12.427257           6.325286     13.514134   
min     1318.000000     19.000000           0.000000      1.000000   
25%     3360.000000     31.000000           5.000000     16.000000   
50%     5031.000000     42.000000          10.000000     28.000000   
75%     6335.000000     53.000000          15.000000     42.000000   
max     8336.000000     65.000000          26.000000     46.000000   

        STATUS_YEAR  
count  49653.000000  
mean    2010.612612  
std        2.845577  
min     2006.000000  
25%     2008.000000  
50%     2011.000000  
75%     2013.000000  
max     2015.000000

**Observations:** no outliers observed. Only _age_ and _length of service_ are true quantitative variables. The mean of _length of service_ is less than the median, but that is unsurprising given that it's less likely now for employees to remain at companies for significant periods of time. 

In [9]:
#look at the shape and size of our dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49653 entries, 0 to 49652
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   EmployeeID           49653 non-null  int64 
 1   recorddate_key       49653 non-null  object
 2   birthdate_key        49653 non-null  object
 3   orighiredate_key     49653 non-null  object
 4   terminationdate_key  49653 non-null  object
 5   age                  49653 non-null  int64 
 6   length_of_service    49653 non-null  int64 
 7   country              49653 non-null  object
 8   department_name      49653 non-null  object
 9   job_title            49653 non-null  object
 10  store_name           49653 non-null  int64 
 11  gender_short         49653 non-null  object
 12  gender_full          49653 non-null  object
 13  termreason_desc      49653 non-null  object
 14  termtype_desc        49653 non-null  object
 15  STATUS_YEAR          49653 non-null  int64 
 16  STAT

In [10]:
df['terminationdate_key'].value_counts()

01/01/1900    42450
12/30/2014     1079
12/30/2015      674
12/30/2010       25
11/11/2012       21
              ...  
11/13/2006        1
10/31/2006        1
10/30/2006        1
10/03/2006        1
9/14/2013         1
Name: terminationdate_key, Length: 1055, dtype: int64

**Overall observations:** 
 1. The dataset is of a reasonable size, only 6.8+ MB. 
 2. We have a mix of numeric and string variables, although all numeric values are integers. That being said, we appear to have some date keys which are returned as string values, however, these may need to be changed to datetime values if we need them for our analyses. 
 3. We have no null values. 
 4. There are a couple of variables which contain similar information, for example, *orighiredate_key* and *length_of_service*, and *gender_short* and *gender_full*. It is likely that I will only need to keep *length_of_service* and *gender_short*. 
 5. *store_name* is given as a numeric variable. 
 6. *terminationdate_key*, *termreason_desc*, *termtype_desc*, and *STATUS* all provide different details about the termination status of an employee. *terminationdate_key* has 42450 keys categorised to 1/1/1900 which I believe is being used an indication that the employee has not yet been terminated. 


# 3. Addressing our dataset

## A. Wrangling

### Approach to data wrangling 
 
   1. Check data types of columns to ensure they are correct and appropriate for the type of data collected
   2. Amend any spelling and syntactical errors in the column titles. This will include use of capital and lowercase letters, and use of underscores and spaces. 
   3. Investigating naming conventions
   4. Look at whether we need to drop any columns 

##### 1. Amending datatypes

I will be amending the datatypes for some of the numeric variables in order to avoid any memory issues later on in this project.

In [11]:
df.dtypes

EmployeeID              int64
recorddate_key         object
birthdate_key          object
orighiredate_key       object
terminationdate_key    object
age                     int64
length_of_service       int64
country                object
department_name        object
job_title              object
store_name              int64
gender_short           object
gender_full            object
termreason_desc        object
termtype_desc          object
STATUS_YEAR             int64
STATUS                 object
BUSINESS_UNIT          object
dtype: object

In [12]:
#amending datatype for user_id, age, n_dependents, and income
df['EmployeeID']=df['EmployeeID'].astype('int16')
df['age'] =df['age'].astype('int8')
df['length_of_service'] = df['length_of_service'].astype('int16')
df['store_name']=df['store_name'].astype('int8')
df['STATUS_YEAR']=df['STATUS_YEAR'].astype('int16')

In [13]:
df.dtypes

EmployeeID              int16
recorddate_key         object
birthdate_key          object
orighiredate_key       object
terminationdate_key    object
age                      int8
length_of_service       int16
country                object
department_name        object
job_title              object
store_name               int8
gender_short           object
gender_full            object
termreason_desc        object
termtype_desc          object
STATUS_YEAR             int16
STATUS                 object
BUSINESS_UNIT          object
dtype: object

##### 2. Amending spelling and syntax errors in columns names

In [14]:
#amending EmployeeID, store_name, STATUS_YEAR, STATUS_YEAR, STATUS, and BUSINESS_UNIT columns
df.rename(columns = {'EmployeeID':'employee_id', 'store_name':'store_number', 'STATUS_YEAR':'year_of_record', 'gender_short':'gender', 'STATUS':'employment_status', 'BUSINESS_UNIT':'business_unit'}, inplace = True)

In [15]:
df.columns

Index(['employee_id', 'recorddate_key', 'birthdate_key', 'orighiredate_key',
       'terminationdate_key', 'age', 'length_of_service', 'country',
       'department_name', 'job_title', 'store_number', 'gender', 'gender_full',
       'termreason_desc', 'termtype_desc', 'year_of_record',
       'employment_status', 'business_unit'],
      dtype='object')

In [16]:
#removing the space in the column titles  
df.columns.str.strip()

Index(['employee_id', 'recorddate_key', 'birthdate_key', 'orighiredate_key',
       'terminationdate_key', 'age', 'length_of_service', 'country',
       'department_name', 'job_title', 'store_number', 'gender', 'gender_full',
       'termreason_desc', 'termtype_desc', 'year_of_record',
       'employment_status', 'business_unit'],
      dtype='object')

##### 3. Investigating naming conventions

Looking into whether there are incorrect spellings in some of the columns

In [17]:
df['department_name'].value_counts()

Meats                     10269
Dairy                      8599
Produce                    8515
Bakery                     8381
Customer Service           7122
Processed Foods            5911
Store Management            271
Executive                   100
Recruitment                  72
HR Technology                64
Accounting                   59
Employee Records             44
Accounts Receiveable         39
Labor Relations              34
Accounts Payable             34
Training                     30
Compensation                 24
Audit                        24
Investment                   24
Information Technology       20
Legal                        17
Name: department_name, dtype: int64

**Observations:** data returned has no outliers or mispelt entries

In [18]:
df['gender'].value_counts()

F    25898
M    23755
Name: gender, dtype: int64

**Observations:** data returned has no outliers or mispelt entries

In [19]:
#looking at gender_full variable
df['gender_full'].value_counts()

Female    25898
Male      23755
Name: gender_full, dtype: int64

**Observations:** data returned no outliers or mispelt entries

In [20]:
#looking at termination reason variable
df['termreason_desc'].value_counts()

Not Applicable    48168
Retirement          885
Resignaton          385
Layoff              215
Name: termreason_desc, dtype: int64

**Observations:** data returned no outliers or mispelt entries

In [21]:
#looking at termination type variable
df['termtype_desc'].value_counts()

Not Applicable    48168
Voluntary          1270
Involuntary         215
Name: termtype_desc, dtype: int64

**Observations:** data returned no outliers or mispelt entries

In [22]:
#looking at employment status variable
df['employment_status'].value_counts()

ACTIVE        48168
TERMINATED     1485
Name: employment_status, dtype: int64

**Observations:** data returned no outliers or mispelt entries, however, this variable can be categorised to **Boolean** for our analysis. 

In [23]:
#looking into business unit variable 
df['business_unit'].value_counts()

STORES        49068
HEADOFFICE      585
Name: business_unit, dtype: int64

**Observations:** data returned no outliers or mispelt entries

In [24]:
#looking at employment status
df['employment_status'].value_counts()

ACTIVE        48168
TERMINATED     1485
Name: employment_status, dtype: int64

In [25]:
df['employment_status'].value_counts(normalize = True).mul(100).round(1).astype(str) + '%'

ACTIVE        97.0%
TERMINATED     3.0%
Name: employment_status, dtype: object

**Observations:** data returned no outliers or mispelt entries

##### 4. Dropping columns

Investigating whether we need to drop any columns from our dataset. 

In [26]:
df.head()

employee_id   recorddate_key birthdate_key orighiredate_key  \
0         1318  12/31/2006 0:00    01/03/1954        8/28/1989   
1         1318  12/31/2007 0:00    01/03/1954        8/28/1989   
2         1318  12/31/2008 0:00    01/03/1954        8/28/1989   
3         1318  12/31/2009 0:00    01/03/1954        8/28/1989   
4         1318  12/31/2010 0:00    01/03/1954        8/28/1989   

  terminationdate_key  age  length_of_service  country department_name  \
0          01/01/1900   52                 17  Germany       Executive   
1          01/01/1900   53                 18  Germany       Executive   
2          01/01/1900   54                 19  Germany       Executive   
3          01/01/1900   55                 20  Germany       Executive   
4          01/01/1900   56                 21  Germany       Executive   

  job_title  store_number gender gender_full termreason_desc   termtype_desc  \
0       CEO            35      M        Male  Not Applicable  Not Applicable   
1       CEO            35      M        Male  Not Applicable  Not Applicable   
2       CEO            35      M        Male  Not Applicable  Not Applicable   
3       CEO            35      M        Male  Not Applicable  Not Applicable   
4       CEO            35      M        Male  Not Applicable  Not Applicable   

   year_of_record employment_status business_unit  
0            2006            ACTIVE    HEADOFFICE  
1            2007            ACTIVE    HEADOFFICE  
2            2008            ACTIVE    HEADOFFICE  
3            2009            ACTIVE    HEADOFFICE  
4            2010            ACTIVE    HEADOFFICE

I am going to remove four columns before conducting our analyses:

    1. 'employee_id' is not needed for our analyses. 
    2. 'recorddate_key' captures the date and time information was captured about the employee, and would not be useful for our analysis.
    3. 'birthdate_key' captures the date of birthday for the employee, but we also have a variable *age* which is more useful for our analysis. 
    4. 'gender_full' details the employees gender written as a full word, but we also have the variable *gender_short* which captures the same information with less memory. 
    5. 'orighiredate_key' details the date the employee was hired, however, we also have the length_of_service which captures similar information in an easier format. 

In [27]:
#drop recorddate_key, birthdate_key, and gender_full columns 
df = df.drop(['employee_id','recorddate_key', 'birthdate_key', 'gender_full', 'orighiredate_key'], axis = 1)

In [28]:
#checking output 
df.head()

terminationdate_key  age  length_of_service  country department_name  \
0          01/01/1900   52                 17  Germany       Executive   
1          01/01/1900   53                 18  Germany       Executive   
2          01/01/1900   54                 19  Germany       Executive   
3          01/01/1900   55                 20  Germany       Executive   
4          01/01/1900   56                 21  Germany       Executive   

  job_title  store_number gender termreason_desc   termtype_desc  \
0       CEO            35      M  Not Applicable  Not Applicable   
1       CEO            35      M  Not Applicable  Not Applicable   
2       CEO            35      M  Not Applicable  Not Applicable   
3       CEO            35      M  Not Applicable  Not Applicable   
4       CEO            35      M  Not Applicable  Not Applicable   

   year_of_record employment_status business_unit  
0            2006            ACTIVE    HEADOFFICE  
1            2007            ACTIVE    HEADOFFICE  
2            2008            ACTIVE    HEADOFFICE  
3            2009            ACTIVE    HEADOFFICE  
4            2010            ACTIVE    HEADOFFICE

In [29]:
df.shape

(49653, 13)

**Observations:** The four columns have been removed. 

## B. Consistency checks

### Approach to consistency checks 

1. Checking for duplicate values
2. Checking for mixed-type data columns 
3. Checking for missing values 

##### 1. Checking for duplicate values 

In [30]:
#creating new data frame just for duplicate values
df_dups = df[df.duplicated()]

In [31]:
#printing duplicate values
df_dups

terminationdate_key  age  length_of_service     country  \
430             6/16/2009   57                 17     Germany   
431             6/16/2009   58                 18     Germany   
432             6/16/2009   59                 19     Germany   
954            12/30/2010   56                 16      France   
955            12/30/2010   57                 17      France   
...                   ...  ...                ...         ...   
49545          12/30/2014   25                  3       Italy   
49586          02/12/2014   21                  1     Germany   
49596          12/30/2014   22                  2       Italy   
49599          12/30/2014   22                  2  Montenegro   
49600          12/30/2014   22                  2       Italy   

              department_name        job_title  store_number gender  \
430    Information Technology  Systems Analyst            35      M   
431    Information Technology  Systems Analyst            35      M   
432    Information Technology  Systems Analyst            35      M   
954                     Meats      Meat Cutter            37      M   
955                     Meats      Meat Cutter            37      M   
...                       ...              ...           ...    ...   
49545         Processed Foods    Shelf Stocker            20      F   
49586        Customer Service          Cashier            44      M   
49596        Customer Service          Cashier            20      M   
49599        Customer Service          Cashier            11      F   
49600        Customer Service          Cashier            20      M   

      termreason_desc   termtype_desc  year_of_record employment_status  \
430    Not Applicable  Not Applicable            2006            ACTIVE   
431    Not Applicable  Not Applicable            2007            ACTIVE   
432    Not Applicable  Not Applicable            2008            ACTIVE   
954    Not Applicable  Not Applicable            2006            ACTIVE   
955    Not Applicable  Not Applicable            2007            ACTIVE   
...               ...             ...             ...               ...   
49545          Layoff     Involuntary            2014        TERMINATED   
49586      Resignaton       Voluntary            2014        TERMINATED   
49596          Layoff     Involuntary            2014        TERMINATED   
49599          Layoff     Involuntary            2014        TERMINATED   
49600          Layoff     Involuntary            2014        TERMINATED   

      business_unit  
430      HEADOFFICE  
431      HEADOFFICE  
432      HEADOFFICE  
954          STORES  
955          STORES  
...             ...  
49545        STORES  
49586        STORES  
49596        STORES  
49599        STORES  
49600        STORES  

[24396 rows x 13 columns]

**Observations:** No duplicate values found 

##### 2. Checking for mixed-type data

In [32]:
#Checking for mixed-type data in the dataframe
for col in df.columns.tolist():
    weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df[weird]) > 0:
        print (col)

**Observations:** No mixed-data returned

##### 3. Checking for missing values

In [33]:
#checking for missing values in the datafram
df.isnull().sum()

terminationdate_key    0
age                    0
length_of_service      0
country                0
department_name        0
job_title              0
store_number           0
gender                 0
termreason_desc        0
termtype_desc          0
year_of_record         0
employment_status      0
business_unit          0
dtype: int64

**Observations:** No missing data observed

# 4. Merging population data

### A. Adding in town or city population

At the moment, the dataset only contains the city or town where the employee is based but I want to understand the size of these locations to see if they have an impact on attrition. 

I used the census information held on the Government website Statistic Canada to obtain the population information for each of the towns from 2015. I created a csv containing this information which I will upload. 

In [34]:
#importing csv containing location and population information
df_pop = pd.read_csv(os.path.join(path, '02_data', 'Original_data', 'EuPopulations.csv'), index_col=False)

In [35]:
#checking output
print('EU Populations')
df_pop

EU Populations


#                   country  population  Yearly  Change  Net  Change  \
0    1                    Russia   145934462          0.0004        62206   
1    2                   Germany    83783942          0.0032       266897   
2    3            United Kingdom    67886011          0.0053       355839   
3    4                    France    65273511          0.0022       143783   
4    5                     Italy    60461826         -0.0015       -88249   
5    6                     Spain    46754778          0.0004        18002   
6    7                   Ukraine    43733762         -0.0059      -259876   
7    8                    Poland    37846611         -0.0011       -41157   
8    9                   Romania    19237691         -0.0066      -126866   
9   10               Netherlands    17134872          0.0022        37742   
10  11                   Belgium    11589623          0.0044        50295   
11  12  Czech Republic (Czechia)    10708981          0.0018        19772   
12  13                    Greece    10423054         -0.0048       -50401   
13  14                  Portugal    10196709         -0.0029       -29478   
14  15                    Sweden    10099265          0.0063        62886   
15  16                   Hungary     9660351         -0.0025       -24328   
16  17                   Belarus     9449323         -0.0003        -3088   
17  18                   Austria     9006398          0.0057        51296   
18  19                    Serbia     8737371         -0.0040       -34864   
19  20               Switzerland     8654622          0.0074        63257   
20  21                  Bulgaria     6948445         -0.0074       -51674   
21  22                   Denmark     5792202          0.0035        20326   
22  23                   Finland     5540720          0.0015         8564   
23  24                  Slovakia     5459642          0.0005         2629   
24  25                    Norway     5421241          0.0079        42384   
25  26                   Ireland     4937786          0.0113        55291   
26  27                   Croatia     4105267         -0.0061       -25037   
27  28                   Moldova     4033963         -0.0023        -9300   
28  29    Bosnia and Herzegovina     3280819         -0.0061       -20181   
29  30                   Albania     2877797         -0.0011        -3120   
30  31                 Lithuania     2722289         -0.0135       -37338   
31  32           North Macedonia     2083374          0.0000          -85   
32  33                  Slovenia     2078938          0.0001          284   
33  34                    Latvia     1886198         -0.0108       -20545   
34  35                   Estonia     1326535          0.0007          887   
35  36                Montenegro      628066          0.0001           79   
36  37                Luxembourg      625978          0.0166        10249   
37  38                     Malta      441543          0.0027         1171   
38  39                   Iceland      341243          0.0065         2212   
39  40           Channel Islands      173863          0.0093         1604   
40  41               Isle of Man       85033          0.0053          449   
41  42                   Andorra       77265          0.0016          123   
42  43            Faeroe Islands       48863          0.0038          185   
43  44                    Monaco       39242          0.0071          278   
44  45             Liechtenstein       38128          0.0029          109   
45  46                San Marino       33931          0.0021           71   
46  47                 Gibraltar       33691         -0.0003          -10   
47  48                  Holy See         801          0.0025            2   

    Density  (P/Km²)  
0                  9  
1                240  
2                281  
3                119  
4                206  
5                 94  
6                 75  
7                124  
8                 84  
9

In [36]:
df_pop.shape

(48, 6)

In [37]:
#checking datatypes
df_pop.dtypes

#                     int64
country              object
population            int64
Yearly  Change      float64
Net  Change           int64
Density  (P/Km²)      int64
dtype: object

In [38]:
#removing the space in the column titles  
df_pop.columns.str.strip()

Index(['#', 'country', 'population', 'Yearly  Change', 'Net  Change',
       'Density  (P/Km²)'],
      dtype='object')

In [39]:
#dropping columns not needed for our analyses prior to merging
df_pop = df_pop.drop(columns = ['#','Yearly  Change', 'Net  Change', 'Density  (P/Km²)'])

In [40]:
#renaming "country column" for merging
df_pop.head()

country  population
0          Russia   145934462
1         Germany    83783942
2  United Kingdom    67886011
3          France    65273511
4           Italy    60461826

In [41]:
#merging df_ords and df_ords_prior
df_merge = df.merge(df_pop, on = 'country', indicator = True)

In [42]:
#checking output
df_merge

terminationdate_key  age  length_of_service  country department_name  \
0              01/01/1900   52                 17  Germany       Executive   
1              01/01/1900   53                 18  Germany       Executive   
2              01/01/1900   54                 19  Germany       Executive   
3              01/01/1900   55                 20  Germany       Executive   
4              01/01/1900   56                 21  Germany       Executive   
...                   ...  ...                ...      ...             ...   
49639           9/13/2013   65                 13  Iceland          Bakery   
49640          12/30/2015   31                  7  Iceland           Dairy   
49641          12/30/2015   31                  7  Iceland           Dairy   
49642          12/30/2015   29                  6  Iceland           Dairy   
49643          12/30/2015   23                  3  Iceland           Dairy   

          job_title  store_number gender termreason_desc   termtype_desc  \
0               CEO            35      M  Not Applicable  Not Applicable   
1               CEO            35      M  Not Applicable  Not Applicable   
2               CEO            35      M  Not Applicable  Not Applicable   
3               CEO            35      M  Not Applicable  Not Applicable   
4               CEO            35      M  Not Applicable  Not Applicable   
...             ...           ...    ...             ...             ...   
49639         Baker            24      F      Retirement       Voluntary   
49640  Dairy Person            24      M          Layoff     Involuntary   
49641  Dairy Person            24      M          Layoff     Involuntary   
49642  Dairy Person            24      M          Layoff     Involuntary   
49643  Dairy Person            24      F          Layoff     Involuntary   

       year_of_record employment_status business_unit  population _merge  
0                2006            ACTIVE    HEADOFFICE    83783942   both  
1                2007            ACTIVE    HEADOFFICE    83783942   both  
2                2008            ACTIVE    HEADOFFICE    83783942   both  
3                2009            ACTIVE    HEADOFFICE    83783942   both  
4                2010            ACTIVE    HEADOFFICE    83783942   both  
...               ...               ...           ...         ...    ...  
49639            2013        TERMINATED        STORES      341243   both  
49640            2015        TERMINATED        STORES      341243   both  
49641            2015        TERMINATED        STORES      341243   both  
49642            2015        TERMINATED        STORES      341243   both  
49643            2015        TERMINATED        STORES      341243   both  

[49644 rows x 15 columns]

In [43]:
df_merge['_merge'].value_counts()

both          49644
left_only         0
right_only        0
Name: _merge, dtype: int64

In [44]:
df_merge.shape

(49644, 15)

### B. Creating a population flag 

I want to understand whether popualtion size has an impact on attrition, so I'm going to group the city locations into groups. Statistics Canada defines population areas in the following sizes:
- small countries, with a population of up to 9,999,999;
- medium countries, with a population of between 10,000,000 and 49,999,999;
- large countries, consisting of a population of 50,000,000 and over.

In [45]:
#creating pop_flag variable
df_merge.loc[df_merge['population'] > 50000000, 'pop_flag'] = 'large_country'
df_merge.loc[(df_merge['population'] <= 49999999) & (df_merge['population'] > 10000000), 'pop_flag'] = 'medium_country'
df_merge.loc[df_merge['population'] < 9999999, 'pop_flag'] = 'small_country'

In [46]:
#checking output 
df_merge['pop_flag'].value_counts()

large_country     21121
medium_country    14650
small_country     13873
Name: pop_flag, dtype: int64

In [47]:
#checking output as percentage 
df_merge['pop_flag'].value_counts(normalize = True).mul(100).round(1).astype(str) + '%'

large_country     42.5%
medium_country    29.5%
small_country     27.9%
Name: pop_flag, dtype: object

In [48]:
#checking dataframe size - should have 20 columns now 
df_merge.shape

(49644, 16)

### C. Creating job groupings

Another key variable we have is job title and a potentially interesting part of analyses could be whether job seniority impacts attrition. Unfortunately, there are over 40 different job titles within the dataset. I will therefore group the different jobs into four bands:

    1. C-Suite: CEO, VP, Chief Information Officer 
    2. Director: anyone with Director in their title
    3. Manager: anyone with Manager in their title
    4. Associate: everyone else including Legal Counsel, Analysts, Clerks, Executive Assistants etc. 

In [49]:
#understanding how many jobs there currently are
df_merge['job_title'].nunique()

47

In [50]:
df_merge['job_title'].value_counts()

Meat Cutter                        9984
Dairy Person                       8581
Produce Clerk                      8237
Baker                              8096
Cashier                            6816
Shelf Stocker                      5622
Customer Service Manager            306
Processed Foods Manager             289
Bakery Manager                      285
Meats Manager                       285
Produce Manager                     278
Store Manager                       271
Recruiter                            62
HRIS Analyst                         55
Accounting Clerk                     50
Benefits Admin                       35
Accounts Receiveable Clerk           30
Labor Relations Analyst              30
Trainer                              26
Accounts Payable Clerk               25
Investment Analyst                   20
Auditor                              20
Systems Analyst                      20
Compensation Analyst                 20
Corporate Lawyer                     17


In [51]:
#grouping job titles into four categories
c_suite = ['CEO', 'CHief Information Officer', 'VP Stores', 'VP Human Resources', 'VP Finance']
director = ['Director, HR Technology', 'Director, Employee Records', 'Director, Accounts Receivable', 'Director, Accounts Payable', 'Director, Accounting', 'Director, Investments', 'Director, Labor Relations', 'Director, Compensation', 'Director, Audit', 'Director, Training', 'Director, Recruitment']
manager = ['Dairy Manager', 'Store Manager', 'Produce Manager', 'Bakery Manager', 'Meats Manager', 'Processed Foods Manager', 'Customer Service Manager']
associate = ['Meat Cutter', 'Dairy Person', 'Produce Clerk', 'Baker', 'Cashier', 'Shelf Stocker', 'Recruiter', 'HRIS Analyst', 'Accounting Clerk', 'Benefits Admin', 'Accounts Receiveable Clerk', 'Labor Relations Analyst', 'Trainer', 'Accounts Payable Clerk', 'Investment Analyst', 'Auditor', 'Systems Analyst', 'Compensation Analyst', 'Corporate Lawyer', 'Legal Counsel', 'Exec Assistant, VP Stores', 'Exec Assistant, Legal Counsel', 'Exec Assistant, Human Resources', 'Exec Assistant, Finance']

In [52]:
#checking the output 
total = len(c_suite) + len(director) + len(manager) + len(associate)
print('Total jobs categorised:', total, 'out of 47')

Total jobs categorised: 47 out of 47


In [53]:
#creating new duplicate job_title column called 'seniority' in case we need job_title information further in our analyses
df_merge['seniority'] = df_merge.loc[:, 'job_title']

In [54]:
#defining a function which groups our job_titles into our four categories and mapping the results to the seniority column 
def titleChange(job):
    if job in c_suite: return 'c_suite'
    if job in director: return 'director'
    if job in manager: return 'manager'
    if job in associate: return 'associate'
    else: return 'new_role'

df_merge['seniority'] = df_merge['seniority'].map(titleChange)

In [55]:
df_merge['seniority'].value_counts()

associate    47796
manager       1723
director        75
c_suite         50
Name: seniority, dtype: int64

In [56]:
#checking output 
df_merge.head()

terminationdate_key  age  length_of_service  country department_name  \
0          01/01/1900   52                 17  Germany       Executive   
1          01/01/1900   53                 18  Germany       Executive   
2          01/01/1900   54                 19  Germany       Executive   
3          01/01/1900   55                 20  Germany       Executive   
4          01/01/1900   56                 21  Germany       Executive   

  job_title  store_number gender termreason_desc   termtype_desc  \
0       CEO            35      M  Not Applicable  Not Applicable   
1       CEO            35      M  Not Applicable  Not Applicable   
2       CEO            35      M  Not Applicable  Not Applicable   
3       CEO            35      M  Not Applicable  Not Applicable   
4       CEO            35      M  Not Applicable  Not Applicable   

   year_of_record employment_status business_unit  population _merge  \
0            2006            ACTIVE    HEADOFFICE    83783942   both   
1            2007            ACTIVE    HEADOFFICE    83783942   both   
2            2008            ACTIVE    HEADOFFICE    83783942   both   
3            2009            ACTIVE    HEADOFFICE    83783942   both   
4            2010            ACTIVE    HEADOFFICE    83783942   both   

        pop_flag seniority  
0  large_country   c_suite  
1  large_country   c_suite  
2  large_country   c_suite  
3  large_country   c_suite  
4  large_country   c_suite

In [57]:
#confirming that dataframe has the same number of rows but one more column
df_merge.shape

(49644, 17)

### D. Creating age groupings

The dataset includes a variable for age, but it might also be interesting to aggregate age into three groups. 

In [58]:
#understanding the age range of employees
df_merge['age'].nunique()

47

In [59]:
df_merge['age'].describe()

count    49644.000000
mean        42.077411
std         12.428304
min         19.000000
25%         31.000000
50%         42.000000
75%         53.000000
max         65.000000
Name: age, dtype: float64

**Observations:** There are 47 different ages within our dataset. The minimum age is 19 and the maximum is 65, which would be retirement age. I am going to group employees into four groups: 

1. Young adults: 19 - 34
2. Mid adults: 35 - 49
3. Older adults: 50 - 65 

In [60]:
#creating age_flag variable
df_merge.loc[df_merge['age'] >= 50, 'age_flag'] = 'older_adults'
df_merge.loc[(df_merge['age'] < 50) & (df_merge['age'] >= 35), 'age_flag'] = 'mid_adults'
df_merge.loc[df_merge['age'] < 35, 'age_flag'] = 'young_adults'

In [61]:
#checking output 
df_merge['age_flag'].value_counts()

mid_adults      17378
older_adults    16154
young_adults    16112
Name: age_flag, dtype: int64

In [62]:
#checking output as percentage 
df_merge['age_flag'].value_counts(normalize = True).mul(100).round(1).astype(str) + '%'

mid_adults      35.0%
older_adults    32.5%
young_adults    32.5%
Name: age_flag, dtype: object

In [63]:
#checking dataframe size - should have 19 columns now 
df_merge.shape

(49644, 18)

# 5. Converting categorical variables to numeric

Any future machine learning models require categorical variables to be coded as numerical values. We will start by understanding which variables will need to be encoded. 

#### Understanding which variables need to be converted

In [64]:
#reminder of the dataset
df_merge.head()

terminationdate_key  age  length_of_service  country department_name  \
0          01/01/1900   52                 17  Germany       Executive   
1          01/01/1900   53                 18  Germany       Executive   
2          01/01/1900   54                 19  Germany       Executive   
3          01/01/1900   55                 20  Germany       Executive   
4          01/01/1900   56                 21  Germany       Executive   

  job_title  store_number gender termreason_desc   termtype_desc  \
0       CEO            35      M  Not Applicable  Not Applicable   
1       CEO            35      M  Not Applicable  Not Applicable   
2       CEO            35      M  Not Applicable  Not Applicable   
3       CEO            35      M  Not Applicable  Not Applicable   
4       CEO            35      M  Not Applicable  Not Applicable   

   year_of_record employment_status business_unit  population _merge  \
0            2006            ACTIVE    HEADOFFICE    83783942   both   
1            2007            ACTIVE    HEADOFFICE    83783942   both   
2            2008            ACTIVE    HEADOFFICE    83783942   both   
3            2009            ACTIVE    HEADOFFICE    83783942   both   
4            2010            ACTIVE    HEADOFFICE    83783942   both   

        pop_flag seniority      age_flag  
0  large_country   c_suite  older_adults  
1  large_country   c_suite  older_adults  
2  large_country   c_suite  older_adults  
3  large_country   c_suite  older_adults  
4  large_country   c_suite  older_adults

In [65]:
#creating a dataframe of only categorical variables to be converted to numeric
df_categorical = df_merge[['gender', 'termreason_desc', 'termtype_desc', 'employment_status', 'business_unit', 'pop_flag', 'seniority', 'age_flag']]
df_categorical.head()

gender termreason_desc   termtype_desc employment_status business_unit  \
0      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
1      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
2      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
3      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
4      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   

        pop_flag seniority      age_flag  
0  large_country   c_suite  older_adults  
1  large_country   c_suite  older_adults  
2  large_country   c_suite  older_adults  
3  large_country   c_suite  older_adults  
4  large_country   c_suite  older_adults

In [66]:
#print column names and all unique variables
for column in df_categorical:
    print (f'{column} : {df_categorical[column].unique()}')

gender : ['M' 'F']
termreason_desc : ['Not Applicable' 'Retirement' 'Resignaton' 'Layoff']
termtype_desc : ['Not Applicable' 'Voluntary' 'Involuntary']
employment_status : ['ACTIVE' 'TERMINATED']
business_unit : ['HEADOFFICE' 'STORES']
pop_flag : ['large_country' 'small_country' 'medium_country']
seniority : ['c_suite' 'associate' 'director' 'manager']
age_flag : ['older_adults' 'mid_adults' 'young_adults']


In [67]:
df_merge.head()

terminationdate_key  age  length_of_service  country department_name  \
0          01/01/1900   52                 17  Germany       Executive   
1          01/01/1900   53                 18  Germany       Executive   
2          01/01/1900   54                 19  Germany       Executive   
3          01/01/1900   55                 20  Germany       Executive   
4          01/01/1900   56                 21  Germany       Executive   

  job_title  store_number gender termreason_desc   termtype_desc  \
0       CEO            35      M  Not Applicable  Not Applicable   
1       CEO            35      M  Not Applicable  Not Applicable   
2       CEO            35      M  Not Applicable  Not Applicable   
3       CEO            35      M  Not Applicable  Not Applicable   
4       CEO            35      M  Not Applicable  Not Applicable   

   year_of_record employment_status business_unit  population _merge  \
0            2006            ACTIVE    HEADOFFICE    83783942   both   
1            2007            ACTIVE    HEADOFFICE    83783942   both   
2            2008            ACTIVE    HEADOFFICE    83783942   both   
3            2009            ACTIVE    HEADOFFICE    83783942   both   
4            2010            ACTIVE    HEADOFFICE    83783942   both   

        pop_flag seniority      age_flag  
0  large_country   c_suite  older_adults  
1  large_country   c_suite  older_adults  
2  large_country   c_suite  older_adults  
3  large_country   c_suite  older_adults  
4  large_country   c_suite  older_adults

In [68]:
#creating a dataframe for the remaing variables which I will merge with the encoded variables 
df_original = df_merge[['country', 'department_name','terminationdate_key', 'age', 'length_of_service', 'job_title', 'store_number', 'year_of_record', 'population', 'pop_flag', 'gender', 'seniority', 'age_flag', 'termreason_desc', 'termtype_desc', 'business_unit']]
df_original.head()

country department_name terminationdate_key  age  length_of_service  \
0  Germany       Executive          01/01/1900   52                 17   
1  Germany       Executive          01/01/1900   53                 18   
2  Germany       Executive          01/01/1900   54                 19   
3  Germany       Executive          01/01/1900   55                 20   
4  Germany       Executive          01/01/1900   56                 21   

  job_title  store_number  year_of_record  population       pop_flag gender  \
0       CEO            35            2006    83783942  large_country      M   
1       CEO            35            2007    83783942  large_country      M   
2       CEO            35            2008    83783942  large_country      M   
3       CEO            35            2009    83783942  large_country      M   
4       CEO            35            2010    83783942  large_country      M   

  seniority      age_flag termreason_desc   termtype_desc business_unit  
0   c_suite  older_adults  Not Applicable  Not Applicable    HEADOFFICE  
1   c_suite  older_adults  Not Applicable  Not Applicable    HEADOFFICE  
2   c_suite  older_adults  Not Applicable  Not Applicable    HEADOFFICE  
3   c_suite  older_adults  Not Applicable  Not Applicable    HEADOFFICE  
4   c_suite  older_adults  Not Applicable  Not Applicable    HEADOFFICE

Once I have encoded my variables, I will merge them with the orginal numeric values. 

### A. Variables with two distinct values

#### a. Gender

In [69]:
#confirming number of unique values prior to encoding so we can check results 
df_categorical['gender'].nunique()

2

In [70]:
#defining department_name encoder object 
gender_encoder = LabelEncoder()

In [71]:
#fit the object into the data
gender_encoder.fit(df_categorical['gender'])

LabelEncoder()

In [72]:
#convert into numerical format
gender_values = gender_encoder.transform(df_categorical['gender'])

In [73]:
print("Before Encoding:", list(df_categorical['gender'][-10:]))
print("After Encoding:", gender_values[-10:])
print("The inverse from the encoding result:", gender_encoder.inverse_transform(gender_values[-10:]))

Before Encoding: ['F', 'F', 'F', 'F', 'M', 'F', 'M', 'M', 'M', 'F']
After Encoding: [0 0 0 0 1 0 1 1 1 0]
The inverse from the encoding result: ['F' 'F' 'F' 'F' 'M' 'F' 'M' 'M' 'M' 'F']


In [74]:
#checking output
df_categorical.head()

gender termreason_desc   termtype_desc employment_status business_unit  \
0      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
1      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
2      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
3      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
4      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   

        pop_flag seniority      age_flag  
0  large_country   c_suite  older_adults  
1  large_country   c_suite  older_adults  
2  large_country   c_suite  older_adults  
3  large_country   c_suite  older_adults  
4  large_country   c_suite  older_adults

In [75]:
df_categorical.shape

(49644, 8)

**Notes** 
Female has been encoded as *0* and Male as *1*

#### b. Employment status

In [76]:
#confirming number of unique values prior to encoding so we can check results 
df_categorical['employment_status'].nunique()

2

In [77]:
#defining department_name encoder object 
employment_encoder = LabelEncoder()

In [78]:
#encoding values
employment_values = employment_encoder.fit_transform(df_categorical['employment_status'])

print("Before Encoding:", list(df_categorical['employment_status'][:5]))
print("After Encoding:", employment_values[:5])
print("The inverse from the encoding result:", employment_encoder.inverse_transform(employment_values[:5]))

Before Encoding: ['ACTIVE', 'ACTIVE', 'ACTIVE', 'ACTIVE', 'ACTIVE']
After Encoding: [0 0 0 0 0]
The inverse from the encoding result: ['ACTIVE' 'ACTIVE' 'ACTIVE' 'ACTIVE' 'ACTIVE']


In [79]:
#checking output
df_categorical.head()

gender termreason_desc   termtype_desc employment_status business_unit  \
0      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
1      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
2      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
3      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
4      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   

        pop_flag seniority      age_flag  
0  large_country   c_suite  older_adults  
1  large_country   c_suite  older_adults  
2  large_country   c_suite  older_adults  
3  large_country   c_suite  older_adults  
4  large_country   c_suite  older_adults

In [80]:
df_merge.shape

(49644, 18)

In [81]:
df_categorical['employment_status'].nunique()

2

**Notes** 
ACTIVE has been encoded as *0* and TERMINATED as *1*

#### c. Business unit

In [82]:
#confirming number of unique values prior to encoding so we can check results 
df_categorical['business_unit'].nunique()

2

In [83]:
#defining department_name encoder object
business_encoder = LabelEncoder()

In [84]:
#encoding values
business_values = business_encoder.fit_transform(df_categorical['business_unit'])

print("Before Encoding:", list(df_categorical['business_unit'][:5]))
print("After Encoding:", business_values[:5])
print("The inverse from the encoding result:", business_encoder.inverse_transform(business_values[:5]))

Before Encoding: ['HEADOFFICE', 'HEADOFFICE', 'HEADOFFICE', 'HEADOFFICE', 'HEADOFFICE']
After Encoding: [0 0 0 0 0]
The inverse from the encoding result: ['HEADOFFICE' 'HEADOFFICE' 'HEADOFFICE' 'HEADOFFICE' 'HEADOFFICE']


In [85]:
#checking output
df_categorical.head()

gender termreason_desc   termtype_desc employment_status business_unit  \
0      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
1      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
2      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
3      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   
4      M  Not Applicable  Not Applicable            ACTIVE    HEADOFFICE   

        pop_flag seniority      age_flag  
0  large_country   c_suite  older_adults  
1  large_country   c_suite  older_adults  
2  large_country   c_suite  older_adults  
3  large_country   c_suite  older_adults  
4  large_country   c_suite  older_adults

In [86]:
df_categorical.shape

(49644, 8)

In [87]:
df_categorical['employment_status'].nunique()

2

**Notes** 
HEADOFFICE has been encoded as *0* and STORES as *1*

### B. Variables with multiple distinct values

For variables with multiple distinct values we need to ise a 2-dimensional array as the input. 

#### a. Term reason

In [88]:
df_categorical['termreason_desc'].value_counts()

Not Applicable    48160
Retirement          885
Resignaton          385
Layoff              214
Name: termreason_desc, dtype: int64

In [89]:
treason_encoder = OneHotEncoder()
treason_reshaped = np.array(df_categorical['termreason_desc']).reshape(-1, 1)
treason_values = treason_encoder.fit_transform(treason_reshaped)

print(df_categorical['termreason_desc'][:5])
print()
print(treason_values.toarray()[:5])
print()
print(treason_encoder.inverse_transform(treason_values)[:5])

0    Not Applicable
1    Not Applicable
2    Not Applicable
3    Not Applicable
4    Not Applicable
Name: termreason_desc, dtype: object

[[0. 1. 0. 0.]
 [0. 1. 0. 0.]
 [0. 1. 0. 0.]
 [0. 1. 0. 0.]
 [0. 1. 0. 0.]]

[['Not Applicable']
 ['Not Applicable']
 ['Not Applicable']
 ['Not Applicable']
 ['Not Applicable']]


#### b. Term type

In [90]:
df_categorical['termtype_desc'].value_counts()

Not Applicable    48160
Voluntary          1270
Involuntary         214
Name: termtype_desc, dtype: int64

In [91]:
ttype_encoder = OneHotEncoder()
ttype_reshaped = np.array(df_categorical['termtype_desc']).reshape(-1, 1)
ttype_values = ttype_encoder.fit_transform(ttype_reshaped)

print(df_categorical['termtype_desc'][:5])
print()
print(ttype_values.toarray()[:5])
print()
print(ttype_encoder.inverse_transform(ttype_values)[:5])

0    Not Applicable
1    Not Applicable
2    Not Applicable
3    Not Applicable
4    Not Applicable
Name: termtype_desc, dtype: object

[[0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]]

[['Not Applicable']
 ['Not Applicable']
 ['Not Applicable']
 ['Not Applicable']
 ['Not Applicable']]


#### c. Population flag 

In [92]:
df_categorical['pop_flag'].value_counts()

large_country     21121
medium_country    14650
small_country     13873
Name: pop_flag, dtype: int64

In [93]:
popflag_encoder = OneHotEncoder()
popflag_reshaped = np.array(df_categorical['pop_flag']).reshape(-1, 1)
popflag_values = popflag_encoder.fit_transform(popflag_reshaped)

print(df_categorical['pop_flag'][:5])
print()
print(popflag_values.toarray()[:5])
print()
print(popflag_encoder.inverse_transform(popflag_values)[:5])

0    large_country
1    large_country
2    large_country
3    large_country
4    large_country
Name: pop_flag, dtype: object

[[1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]]

[['large_country']
 ['large_country']
 ['large_country']
 ['large_country']
 ['large_country']]


#### d. Seniority

In [94]:
df_categorical['seniority'].value_counts()

associate    47796
manager       1723
director        75
c_suite         50
Name: seniority, dtype: int64

In [95]:
seniority_encoder = OneHotEncoder()
seniority_reshaped = np.array(df_categorical['seniority']).reshape(-1, 1)
seniority_values = seniority_encoder.fit_transform(seniority_reshaped)

print(df_categorical['seniority'][:5])
print()
print(seniority_values.toarray()[:5])
print()
print(seniority_encoder.inverse_transform(seniority_values)[:5])

0    c_suite
1    c_suite
2    c_suite
3    c_suite
4    c_suite
Name: seniority, dtype: object

[[0. 1. 0. 0.]
 [0. 1. 0. 0.]
 [0. 1. 0. 0.]
 [0. 1. 0. 0.]
 [0. 1. 0. 0.]]

[['c_suite']
 ['c_suite']
 ['c_suite']
 ['c_suite']
 ['c_suite']]


#### e. Age flag

In [96]:
df_categorical['age_flag'].value_counts()

mid_adults      17378
older_adults    16154
young_adults    16112
Name: age_flag, dtype: int64

In [97]:
ageflag_encoder = OneHotEncoder()
ageflag_reshaped = np.array(df_categorical['age_flag']).reshape(-1, 1)
ageflag_values = ageflag_encoder.fit_transform(ageflag_reshaped)

print(df_categorical['age_flag'][:5])
print()
print(ageflag_values.toarray()[:5])
print()
print(ageflag_encoder.inverse_transform(ageflag_values)[:5])

0    older_adults
1    older_adults
2    older_adults
3    older_adults
4    older_adults
Name: age_flag, dtype: object

[[0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]]

[['older_adults']
 ['older_adults']
 ['older_adults']
 ['older_adults']
 ['older_adults']]


### C. Creating encoded dataframe

For each column, I will initialize the DataFrame object for creating the dataframe. 

    1. Gender
    2. Employment status
    3. Business unit
    4. Term reason
    5. Term type
    6. Population flag
    7. Seniority
    8. Age flag

In [98]:
#gender
gender = pd.DataFrame(gender_values, columns=['gender'])

In [99]:
#employment status
employmentstatus = pd.DataFrame(employment_values, columns=['employment_status'])

In [100]:
#business unit
businessunit = pd.DataFrame(business_values, columns=['business_unit'])

In [101]:
#termination reason
termreason = pd.DataFrame(treason_values.toarray(), columns=['Not Applicable', 'Retirement', 'Resignaton', 'Layoff'])

In [102]:
#term type 
termtype = pd.DataFrame(ttype_values.toarray(), columns=['Not Applicable', 'Voluntary', 'Involuntary'])

In [103]:
#population flag 
popflag = pd.DataFrame(popflag_values.toarray(), columns=['large_country', 'medium_country', 'small_country'])

In [104]:
#seniority 
seniorityflag = pd.DataFrame(seniority_values.toarray(), columns=['c_suite', 'associate', 'director', 'manager'])

In [105]:
#age flag
ageflag = pd.DataFrame(ageflag_values.toarray(), columns=['older_adults', 'mid_adults', 'young_adults'])

Next we combine all of the categorical columns into one dataframe.

In [106]:
df_categorical_encoded = pd.concat([gender, employmentstatus, businessunit, termreason, termtype, popflag, seniorityflag, ageflag], axis=1)

In [107]:
#checking output
print(df_categorical_encoded.shape)
df_categorical_encoded.head()

(49644, 20)


gender  employment_status  business_unit  Not Applicable  Retirement  \
0       1                  0              0             0.0         1.0   
1       1                  0              0             0.0         1.0   
2       1                  0              0             0.0         1.0   
3       1                  0              0             0.0         1.0   
4       1                  0              0             0.0         1.0   

   Resignaton  Layoff  Not Applicable  Voluntary  Involuntary  large_country  \
0         0.0     0.0             0.0        1.0          0.0            1.0   
1         0.0     0.0             0.0        1.0          0.0            1.0   
2         0.0     0.0             0.0        1.0          0.0            1.0   
3         0.0     0.0             0.0        1.0          0.0            1.0   
4         0.0     0.0             0.0        1.0          0.0            1.0   

   medium_country  small_country  c_suite  associate  director  manager  \
0             0.0            0.0      0.0        1.0       0.0      0.0   
1             0.0            0.0      0.0        1.0       0.0      0.0   
2             0.0            0.0      0.0        1.0       0.0      0.0   
3             0.0            0.0      0.0        1.0       0.0      0.0   
4             0.0            0.0      0.0        1.0       0.0      0.0   

   older_adults  mid_adults  young_adults  
0           0.0         1.0           0.0  
1           0.0         1.0           0.0  
2           0.0         1.0           0.0  
3           0.0         1.0           0.0  
4           0.0         1.0           0.0

Finally, we combine our new encoded dataframe with our the other "original" dataframe we created. As a reminder, this isn't the true original dataframe, rather the one we created which captured variables we didn't need to encode. 

In [108]:
df_new = pd.concat([df_original, df_categorical_encoded], axis=1)

In [109]:
#checking output
print(df_new.shape)
df_new.head()

(49644, 36)


country department_name terminationdate_key  age  length_of_service  \
0  Germany       Executive          01/01/1900   52                 17   
1  Germany       Executive          01/01/1900   53                 18   
2  Germany       Executive          01/01/1900   54                 19   
3  Germany       Executive          01/01/1900   55                 20   
4  Germany       Executive          01/01/1900   56                 21   

  job_title  store_number  year_of_record  population       pop_flag  ...  \
0       CEO            35            2006    83783942  large_country  ...   
1       CEO            35            2007    83783942  large_country  ...   
2       CEO            35            2008    83783942  large_country  ...   
3       CEO            35            2009    83783942  large_country  ...   
4       CEO            35            2010    83783942  large_country  ...   

  large_country medium_country small_country c_suite associate director  \
0           1.0            0.0           0.0     0.0       1.0      0.0   
1           1.0            0.0           0.0     0.0       1.0      0.0   
2           1.0            0.0           0.0     0.0       1.0      0.0   
3           1.0            0.0           0.0     0.0       1.0      0.0   
4           1.0            0.0           0.0     0.0       1.0      0.0   

   manager  older_adults  mid_adults  young_adults  
0      0.0           0.0         1.0           0.0  
1      0.0           0.0         1.0           0.0  
2      0.0           0.0         1.0           0.0  
3      0.0           0.0         1.0           0.0  
4      0.0           0.0         1.0           0.0  

[5 rows x 36 columns]

# 6. Exporting dataset

In [111]:
df_new.to_csv(os.path.join(path, '02_Data', 'Prepared_data', 'encoded_data.csv'))